lib/dirac_twisted_clover.cpp

```diff
@@ -188,9 +210,16 @@ namespace quda {
       DiracWilson::Dslash(out, *tmp2, parity);
```

Comment:
@maddyscientist Had a question about using `DiracWilson::Dslash` for two-flavour fields. I wanted to implement the two-flavour `DiracTwistedCloverPC::Dslash` for the case `dagger && symmetric && !reverse` the same way that you did for a single flavour.  However, this fails because of the call to `DiracWilson::Dslash`.

Now, I can modify `WilsonApply` to use `in.getDslashConstant().volume_4d_cb`, as done for other kernels when dealing with two flavours:

```
 35     inline WilsonApply(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a,
 36                        const ColorSpinorField &x, int parity, bool dagger, const int *comm_override, TimeProfile &profile)
 37     {
 38       constexpr int nDim = 4;
 39       WilsonArg<Float, nColor, nDim, recon> arg(out, in, U, a, x, parity, dagger, comm_override);
 40       Wilson<decltype(arg)> wilson(arg, out, in);
 41 
 42       //dslash::DslashPolicyTune<decltype(wilson)> policy(
 43       //  wilson, const_cast<cudaColorSpinorField *>(static_cast<const cudaColorSpinorField *>(&in)), in.VolumeCB(),
 44       //  in.GhostFaceCB(), profile);
 45       dslash::DslashPolicyTune<decltype(wilson)> policy(
 46         wilson, const_cast<cudaColorSpinorField *>(static_cast<const cudaColorSpinorField *>(&in)),
 47         in.getDslashConstant().volume_4d_cb, in.getDslashConstant().ghostFaceCB, profile);                                             
 48     }
 49   };
                                         
```

but I don't know if it will still do the right thing then (i.e. apply the Wilson Dslash to both flavours). Even if it does do the right thing, I don't know if I'd break correct behaviour for other uses if I do this.

Reply:
I've gone with the following now, although I have to admit that I haven't thought enough whether the DWF-generalisation that is already in the `applyWilson` kernel translates correctly to the two-flavour case.

```
 33   template <typename Float, int nColor, QudaReconstructType recon> struct WilsonApply {
 34 
 35     inline WilsonApply(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a,
 36                        const ColorSpinorField &x, int parity, bool dagger, const int *comm_override, TimeProfile &profile)
 37     {
 38       if( in.Ndim() == 4 ){
 39         constexpr int nDim = 4;
 40         WilsonArg<Float, nColor, nDim, recon> arg(out, in, U, a, x, parity, dagger, comm_override);
 41         Wilson<decltype(arg)> wilson(arg, out, in);
 42 
 43         dslash::DslashPolicyTune<decltype(wilson)> policy(
 44           wilson, const_cast<cudaColorSpinorField *>(static_cast<const cudaColorSpinorField *>(&in)), in.VolumeCB(),
 45           in.GhostFaceCB(), profile);
 46       } else {
 47         // support for two-flavour fields (or fields with general fifth dim, where a Wilson Dslash
 48         // is to be applied to each 4D subspace)
 49         constexpr int nDim = 5;
 50         WilsonArg<Float, nColor, nDim, recon> arg(out, in, U, a, x, parity, dagger, comm_override);
 51         Wilson<decltype(arg)> wilson(arg, out, in);
 52 
 53         dslash::DslashPolicyTune<decltype(wilson)> policy(
 54           wilson, const_cast<cudaColorSpinorField *>(static_cast<const cudaColorSpinorField *>(&in)),
 55           in.VolumeCB(), in.GhostFaceCB(), profile);
 56       }
 57     }
 58   };
```

Reply:
One more change to the strategy here, I hope this doesn't break anything as `WilsonApply` works for two flavour fields now (at least with parallelisation in T, not sure if complications arise for partitionings that require packing). https://github.com/lattice/quda/pull/1121/commits/045992381836aaef5f87340c1d4fcc07e2f75556